pymilvus/bulk_writer/buffer.py

```diff
@@ -290,49 +289,62 @@ def _persist_parquet(self, local_path: str, **kwargs):
-                else:
-                    row.append(str(self._buffer[name][i]))
-            data.append(row)
-
-        rows = [header, *data]
+        data = pd.DataFrame(columns=header)
+        for k, v in self._buffer.items():
+            field_schema = self._fields[k]
+            # When using df.to_csv(arr) to write non-scalar data,
+            # the repr function is used to convert the data to a string.
+            # if the value of arr is [1.0, 2.0], repr(arr) will change with the type of arr:
+            #   when arr is a list, the output is '[1.0, 2.0]'
+            #   when arr is a tuple, the output is '(1.0, 2.0)'
+            #   when arr is a np.array, the output is '[1.0 2.0]'
+            # we needs the output to be '[1.0, 2.0]', consistent with the array format in json
+            # so 1. whether make sure that arr of type
+            #       (BINARY_VECTOR, FLOAT_VECTOR, FLOAT16_VECTOR, BFLOAT16_VECTOR) is a LIST,
+            #    2. or convert arr into a string using json.dumps(arr) first and then add it to df
+            # I choose method 2 here
+            if field_schema.dtype in {
+                DataType.JSON,
+                DataType.ARRAY,
+                DataType.SPARSE_FLOAT_VECTOR,
+                DataType.BINARY_VECTOR,
+                DataType.FLOAT_VECTOR,
+            }:
+                dt = np.dtype("str")
+                arr = []
+                for val in v:
+                    arr.append(json.dumps(val))
+                data[k] = pd.Series(arr, dtype=dt)
+            elif field_schema.dtype in {DataType.FLOAT16_VECTOR, DataType.BFLOAT16_VECTOR}:
+                # special process for float16 vector, the self._buffer stores bytes for
+                # float16 vector, convert the bytes to float list
+                dt = (
+                    np.dtype("bfloat16")
+                    if (field_schema.dtype == DataType.BFLOAT16_VECTOR)
+                    else np.dtype("float16")
+                )
+                arr = []
+                for val in v:
+                    arr.append(json.dumps(np.frombuffer(val, dtype=dt).tolist()))
+                data[k] = pd.Series(arr, dtype=np.dtype("str"))
+            elif field_schema.dtype in {DataType.BOOL}:
+                dt = np.dtype("str")
+                arr = ["true" if x else "false" for x in v]
+                data[k] = pd.Series(arr, dtype=dt)
+            elif field_schema.dtype.name in NUMPY_TYPE_CREATOR:
+                dt = NUMPY_TYPE_CREATOR[field_schema.dtype.name]
+                data[k] = pd.Series(v, dtype=dt)
+            else:
+                data[k] = pd.Series(v)
 
         file_path = Path(local_path + ".csv")
         try:
-            with file_path.open("w", encoding="utf-8") as f:
-                writer = csv.writer(f, delimiter=sep)
-                writer.writerows(rows)
+            data.to_csv(file_path, sep=sep, index=False)
```

Comment:
We should redefine the null value representation here, e.g., "NULL", configurable variable preferred.